### DISCUSSION TITLE
`useAtom(atomFamily(id))` from callback/listener

### DISCUSSION THREAD

**Discussion by T-Damer** [pmndrs/jotai#1706] [Q&A]
Hello, I have `atomFamily` of posts in my project:
```TS
type PostIdSetter = number | Promise<FeedPost[]>

export const postsAtom = atomFamily((feedId: number) =>
  atom(getFeedPosts({ feedId }), (get, set, postIdSetter: PostIdSetter) => {
    if (typeof postIdSetter === 'number') {
      const posts = getFeedPosts({ feedId: postIdSetter })
      set(postsAtom(postIdSetter), posts)
    } else {
      const feedId = get(currentFeedIdAtom)
      set(postsAtom(feedId), postIdSetter)
    }
  })
)
```
Basically it's a `Map` like `{[feedId: number]: Posts[]}` with inital `getFeedPosts`, we use is across the project to cache stuff like this:
```TS
const [posts, setPosts] = useAtom(postsAtom(feedId))
```

I want to listen to the events from the contract on blockchain. It looks like this:
```TS
  useContractEvent({
    contract,
    async listener(
      feedId: number
      newPost: PostData
    ) {
      console.log('listening')
      // There's a spot:
      const [posts, setPosts] = useAtom(postsAtom(feedId)
      console.log('got atom')

      setPosts(Promise.resolve([newPost, ...posts]))
    },
  })
```
My question is how to use `atomFamily` like when we don't know the `key`? The listener logs `listeing`, but not `got atom`
I know that the hook inside the listener is a bad idea

The only way to avoid this is to set listener for every `feedId`, which user sees:
```TS
useEffect(() => {
  useContractEvent(feedId)
  return () => useContractEvent(feedId)
}, [feedId])
```
Should be fine, but I have too many re-renders for some reason :thinking:

**Comment by T-Damer** [comment]
Maybe there's a way to extract whole `atomFamily` mapping and update it separately like `posts[1] = newPosts`?

**Reply by dai-shi** [reply]
Hm, seems like there's a little misconception. atomFamily is the mapping of params and atoms, and it has nothing to do with atom values.

See how atomFamily is implemented and hope to get a better understanding.
```js
// simplified code
const atomFamily = (initializeAtom) => {
  const cache = new Map();
  return (param) => {
    if (!cache.has(param)) {
      cache.set(param, initializeAtom(param));
    }
    return cache.get(param);
  };
};
```
https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035

Creating your own variant of atomFamily is absolutely fine.
